File: codebase/eval/embedder.js
#!/usr/bin/env node
/**
 * embedder.js — Embed 700 chunks bằng Gemini text-embedding-004
 *
 * Input:  codebase/eval/chunks.json
 * Output: codebase/eval/embeddings.json  [{code, text, embedding: [...768]}, ...]
 *
 * Biến môi trường: GEMINI_API_KEY
 * Chạy: node codebase/eval/embedder.js
 *
 * Lưu ý:
 *   - Cache theo code: nếu embeddings.json đã có code đó thì bỏ qua
 *   - Batch 100 chunk / lần gọi (giới hạn Gemini batchEmbedContents)
 *   - taskType: RETRIEVAL_DOCUMENT (cho phía tài liệu)
 */

const fs = require("fs");
const path = require("path");
const https = require("https");
require("./loadenv.js"); // load GEMINI_API_KEY từ .env

const CHUNKS = path.resolve(__dirname, "chunks.json");
const OUT = path.resolve(__dirname, "embeddings.json");
const MODEL = "text-embedding-004";
const BATCH = 100;
const DIM = 768;

const apiKey = process.env.GEMINI_API_KEY;
if (!apiKey) {
  console.error("⚠ GEMINI_API_KEY chưa set — sẽ chạy FALLBACK TF-IDF (không cần API).");
  console.error("  Để có chất lượng tốt hơn, set key rồi chạy lại.");
}

function postBatch(requests) {
  return new Promise((resolve, reject) => {
    const body = JSON.stringify({
      requests: requests.map((r) => ({
        model: `models/${MODEL}`,
        content: { parts: [{ text: r.text }] },
        taskType: "RETRIEVAL_DOCUMENT",
        outputDimensionality: DIM,
      })),
    });
    const url = new URL(
      `https://generativelanguage.googleapis.com/v1beta/models/${MODEL}:batchEmbedContents?key=${apiKey}`
    );
    const req = https.request(
      {
        method: "POST",
        hostname: url.hostname,
        path: url.pathname + url.search,
        headers: { "Content-Type": "application/json", "Content-Length": Buffer.byteLength(body) },
      },
      (res) => {
        let data = "";
        res.on("data", (c) => (data += c));
        res.on("end", () => {
          if (res.statusCode >= 400) {
            reject(new Error(`HTTP ${res.statusCode}: ${data.slice(0, 300)}`));
          } else {
            try {
              resolve(JSON.parse(data));
            } catch (e) {
              reject(e);
            }
          }
        });
      }
    );
    req.on("error", reject);
    req.write(body);
    req.end();
  });
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

/**
 * Fallback TF-IDF — chạy khi không có API key.
 * Cho vector 768 chiều cố định, hầu hết = 0. Đủ để demo flow, không đủ để đo chất lượng.
 */
function buildTfIdfFallback(chunks) {
  const vocab = new Map();
  const docs = chunks.map((c) => {
    const tokens = c.text
      .toLowerCase()
      .replace(/[^\p{L}\p{N}\s]+/gu, " ")
      .split(/\s+/)
      .filter((w) => w.length > 1);
    for (const t of tokens) vocab.set(t, (vocab.get(t) || 0) + 1);
    return tokens;
  });
  // Lấy top 768 từ phổ biến nhất làm chiều vector
  const top = [...vocab.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, DIM)
    .map((e) => e[0]);
  const ix = new Map(top.map((w, i) => [w, i]));

  return chunks.map((c, idx) => {
    const vec = new Array(DIM).fill(0);
    const tokens = docs[idx];
    const tf = new Map();
    for (const t of tokens) tf.set(t, (tf.get(t) || 0) + 1);
    for (const [w, n] of tf) {
      const i = ix.get(w);
      if (i !== undefined) vec[i] = n / tokens.length;
    }
    return { code: c.code, text: c.text, embedding: vec };
  });
}

async function main() {
  const chunks = JSON.parse(fs.readFileSync(CHUNKS, "utf-8"));
  let cache = {};
  if (fs.existsSync(OUT)) {
    try {
      const existing = JSON.parse(fs.readFileSync(OUT, "utf-8"));
      for (const e of existing) cache[e.code] = e;
      console.log(`→ Cache: ${Object.keys(cache).length} embeddings đã có.`);
    } catch {}
  }

  const missing = chunks.filter((c) => !cache[c.code]);
  console.log(`→ Cần embed ${missing.length} đoạn mới (tổng ${chunks.length}).`);

  if (!apiKey) {
    console.log("→ Chạy fallback TF-IDF (chất lượng thấp, chỉ để demo).");
    const fb = buildTfIdfFallback(chunks);
    fs.writeFileSync(OUT, JSON.stringify(fb, null, 0));
    console.log(`→ Đã ghi ${fb.length} embeddings vào ${OUT}`);
    return;
  }

  let done = 0;
  for (let i = 0; i < missing.length; i += BATCH) {
    const batch = missing.slice(i, i + BATCH);
    let attempt = 0;
    while (true) {
      try {
        const res = await postBatch(batch);
        for (let j = 0; j < batch.length; j++) {
          const code = batch[j].code;
          cache[code] = { code, text: batch[j].text, embedding: res.embeddings[j].values };
        }
        done += batch.length;
        console.log(`  ${done}/${missing.length}`);
        break;
      } catch (e) {
        attempt++;
        if (attempt >= 3) {
          console.error(`✗ Lỗi sau 3 lần thử tại batch ${i}: ${e.message}`);
          process.exit(1);
        }
        console.log(`  ! Retry ${attempt}/3 sau lỗi: ${e.message.slice(0, 100)}`);
        await sleep(2000 * attempt);
      }
    }
    await sleep(300); // rate limit nhẹ
  }

  const out = chunks.map((c) => cache[c.code]).filter(Boolean);
  fs.writeFileSync(OUT, JSON.stringify(out));
  console.log(`→ Đã ghi ${out.length} embeddings vào ${OUT}`);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: codebase/eval_py/embedder.py
#!/usr/bin/env python3
"""Port of codebase/eval/embedder.js — Embed chunks with the Jina API."""

import asyncio
import json
import os
import re
import sys
from pathlib import Path

import httpx

EVAL_DIR = Path(__file__).resolve().parents[1] / "eval"
PROJECT_ROOT = EVAL_DIR.parents[1]
CHUNKS = EVAL_DIR / "chunks.json"
OUT = EVAL_DIR / "embeddings.json"
MODEL = "jina-embeddings-v3"
BATCH = 32
DIM = 1024


def load_env() -> None:
    for file in (EVAL_DIR / ".env", PROJECT_ROOT / ".env"):
        if not file.exists():
            continue
        loaded = []
        for raw_line in file.read_text(encoding="utf-8").splitlines():
            line = raw_line.strip()
            if not line or line.startswith("#"):
                continue
            match = re.match(r"^([A-Z_][A-Z0-9_]*)\s*=\s*(.*)$", line, re.IGNORECASE)
            if not match:
                continue
            key, value = match.groups()
            value = value.strip()
            if len(value) >= 2 and value[0] == value[-1] and value[0] in "\"'":
                value = value[1:-1]
            if key not in os.environ:
                os.environ[key] = value
                loaded.append(key)
        print(f"[loadenv] loaded from {file}: {', '.join(loaded)}", file=sys.stderr)
        return
    print("[loadenv] no .env found — relying on process.env", file=sys.stderr)


async def post_batch(client: httpx.AsyncClient, texts: list[str], api_key: str) -> list[list[float]]:
    response = await client.post(
        "https://api.jina.ai/v1/embeddings",
        headers={"Content-Type": "application/json", "Authorization": f"Bearer {api_key}"},
        json={
            "model": MODEL,
            "input": texts,
            "dimensions": DIM,
            "normalized": True,
            "task": "retrieval.passage",
        },
    )
    if response.status_code >= 400:
        raise RuntimeError(f"HTTP {response.status_code}: {response.text[:300]}")
    return [item["embedding"] for item in response.json()["data"]]


async def main() -> None:
    load_env()
    api_key = os.getenv("JINA_API_KEY", "")
    if not api_key:
        print("⚠ JINA_API_KEY chưa set — sẽ chạy FALLBACK TF-IDF (không cần API).", file=sys.stderr)
        raise SystemExit(1)

    chunks = json.loads(CHUNKS.read_text(encoding="utf-8"))
    cache: dict[str, dict] = {}
    if OUT.exists():
        try:
            existing = json.loads(OUT.read_text(encoding="utf-8"))
            for entry in existing:
                cache[entry["code"]] = entry
            print(f"→ Cache: {len(cache)} embeddings đã có.")
        except (OSError, json.JSONDecodeError, KeyError, TypeError):
            pass

    missing = [chunk for chunk in chunks if chunk["code"] not in cache]
    print(f"→ Cần embed {len(missing)} đoạn mới (tổng {len(chunks)}).")

    done = 0
    async with httpx.AsyncClient(timeout=60) as client:
        for start in range(0, len(missing), BATCH):
            batch = missing[start:start + BATCH]
            attempt = 0
            while True:
                try:
                    vecs = await post_batch(client, [chunk["text"] for chunk in batch], api_key)
                    for index, chunk in enumerate(batch):
                        code = chunk["code"]
                        cache[code] = {"code": code, "text": chunk["text"], "embedding": vecs[index]}
                    done += len(batch)
                    print(f"  {done}/{len(missing)}")
                    break
                except Exception as exc:
                    attempt += 1
                    if attempt >= 3:
                        print(f"✗ Lỗi sau 3 lần tại batch {start}: {exc}", file=sys.stderr)
                        raise SystemExit(1)
                    print(f"  ! Retry {attempt}/3: {str(exc)[:100]}")
                    await asyncio.sleep(2 * attempt)
            await asyncio.sleep(0.2)

    output = [cache[chunk["code"]] for chunk in chunks if chunk["code"] in cache]
    OUT.write_text(json.dumps(output, ensure_ascii=False, separators=(",", ":")), encoding="utf-8")
    print(f"→ Đã ghi {len(output)} embeddings vào {OUT}")
    print(f"→ Vector dim: {DIM}")


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        raise SystemExit(1)
